libraries/resource_providers/image_streamer/api300/artifact_bundle.rb

```diff
@@ -0,0 +1,117 @@
+        def load_expected_resources(resource_type, resource_array, add_method)
+          return unless resource_array.any?
+          resource_array.each do |resource_name, read_only = true|
+            resource = load_resource(resource_type, resource_name)
+            # NOTE: Option 2: using this add method to auto generate the add_resource_type using the resource_type
+            add_method = 'add_' + resource_type.to_s.gsub(/([a-z\d])([A-Z])/, '\1_\2').downcase
+            @item.method(add_method).call(resource, read_only)
+          end
+        end
+
+        def create_if_missing
+          load_expected_resources(:BuildPlan, @context.os_build_plans, 'add_build_plan')
+          load_expected_resources(:DeploymentPlan, @context.deployment_plans, 'add_deployment_plan')
+          load_expected_resources(:GoldenImage, @context.golden_images, 'add_golden_image')
+          load_expected_resources(:PlanScript, @context.plan_scripts, 'add_plan_script')
+          super
+        end
+
+        def update_name
+          raise "#{@resource_name} #{@name} does not exist in the appliance. Cannot run update_name action." unless @item.retrieve!
+          raise 'This action requires the new_name field to be specified.' unless @context.new_name
+          return Chef::Log.info("#{@resource_name} '#{@name}' is up to date") if @item[:name] == @context.new_name
+          new_name_in_use = resource_named(:ArtifactBundle).new(@item.client, name: @context.new_name).retrieve!
+          raise "#{@resource_name} name '#{@context.new_name}' is already in use." if new_name_in_use
+          Chef::Log.debug "New name #{@context.new_name} differs from current name'#{@name}' for resource."
+          @context.converge_by "Update #{@resource_name}'s name from '#{@name}' to #{@context.new_name} completed successfully" do
+            @item.update_name(@context.new_name)
+          end
+        end
+
+        def download
+          raise "#{@resource_name} #{@context.file_path} must be specified for this action" unless @context.file_path
+          raise "#{@resource_name} #{@name} does not exist in the appliance. Cannot run download action." unless @item.retrieve!
+          return Chef::Log.info("#{@resource_name} '#{@name}' file already exists in specified location.") if File.file?(@context.file_path)
+          @context.converge_by "Download #{@resource_name}'s '#{@name}.zip' to #{@context.file_path} completed successfully" do
+            @item.download(@context.file_path)
+          end
+        end
+
+        def upload
+          raise "#{@resource_name} '#{@context.file_path}' file does not exist." unless File.file?(@context.file_path)
+          return Chef::Log.info("#{@resource_name} #{@name} already exists in the appliance.") if @item.retrieve!
+          @context.converge_by "Upload of #{@resource_name}'s '#{@name}' completed successfully" do
```

Comment:
You need to check also if the File being uploaded exist before executing the upload.